File: packages/core/src/registry.ts
import { RegistryError } from './errors/base-error.js';

export class Registry<T> {
  private readonly items = new Map<string, T>();
  private readonly label: string;

  constructor(label: string) {
    this.label = label;
  }

  register(name: string, item: T): void {
    if (this.items.has(name)) {
      throw new RegistryError(
        `${this.label} '${name}' is already registered`
      );
    }
    this.items.set(name, item);
  }

  get(name: string): T {
    const item = this.items.get(name);
    if (item === undefined) {
      throw new RegistryError(
        `${this.label} '${name}' is not registered`
      );
    }
    return item;
  }

  tryGet(name: string): T | undefined {
    return this.items.get(name);
  }

  has(name: string): boolean {
    return this.items.has(name);
  }

  getAll(): ReadonlyMap<string, T> {
    return this.items;
  }

  getAllNames(): readonly string[] {
    return [...this.items.keys()];
  }

  unregister(name: string): boolean {
    if (!this.items.has(name)) {
      throw new RegistryError(
        `${this.label} '${name}' is not registered`
      );
    }
    return this.items.delete(name);
  }

  clear(): void {
    this.items.clear();
  }

  get size(): number {
    return this.items.size;
  }
}

File: sdk/python/chameleon/registry.py
"""
Registry — Generic named-item store with runtime register/unregister.
Python port of @cli-agent/core Registry<T>.
"""

from __future__ import annotations

from typing import Generic, TypeVar

T = TypeVar("T")


class RegistryError(Exception):
    pass


class Registry(Generic[T]):
    """Thread-safe registry for named components."""

    def __init__(self, label: str = "Registry") -> None:
        self._label = label
        self._items: dict[str, T] = {}

    def register(self, name: str, item: T) -> None:
        if name in self._items:
            raise RegistryError(f"{self._label} '{name}' is already registered")
        self._items[name] = item

    def get(self, name: str) -> T:
        item = self._items.get(name)
        if item is None:
            raise RegistryError(f"{self._label} '{name}' is not registered")
        return item

    def try_get(self, name: str) -> T | None:
        return self._items.get(name)

    def get_all(self) -> dict[str, T]:
        return dict(self._items)

    def get_all_names(self) -> list[str]:
        return list(self._items.keys())

    def unregister(self, name: str) -> bool:
        if name not in self._items:
            raise RegistryError(f"{self._label} '{name}' is not registered")
        del self._items[name]
        return True

    def __len__(self) -> int:
        return len(self._items)

    def __contains__(self, name: str) -> bool:
        return name in self._items
